splash/html_element.py

```diff
@@ -0,0 +1,283 @@
+        is_textarea = tag == "textarea"
+        is_valid_input = tag == "input" and type in supported
+        is_contenteditable = info["attributes"].get("contenteditable") is not None
+
+        if not (is_textarea or is_valid_input or is_contenteditable):
+            raise DOMError({
+                'message': "Node should be editable"
+            })
+
+    def exists(self):
+        """ Return flag indicating whether element is in DOM """
+        try:
+            exists = self.tab.evaljs("document.contains(%s)" % self.element_js)
+            return bool(exists)
+        except JsError:
+            return False
+
+    @empty_strings_as_none
+    def node_property(self, property_name):
+        """ Return value of the specified property of the element """
+        result = self.tab.evaljs(u"{element}[{property}]".format(
+            element=self.element_js,
+            property=escape_js(property_name)
+        ))
+        return self.return_html_element_if_node(result)
+
+    @empty_strings_as_none
+    def set_node_property(self, property_name, property_value):
+        """ Set value of the specified property of the element """
+        result = self.tab.evaljs(u"{element}[{property}] = {value}".format(
+            element=self.element_js,
+            property=escape_js(property_name),
+            value=escape_js(property_value)
+        ))
+        return self.return_html_element_if_node(result)
+
+    def get_node_style(self, property_name):
+        """ Get value of the style property of the element """
+        return self.tab.evaljs(u"{element}.style[{property}]".format(
+            element=self.element_js,
+            property=escape_js(property_name),
+        ))
+
+    def set_node_style(self, property_name, property_value):
+        """ Set value of the style property of the element """
+        return self.tab.evaljs(u"{element}.style[{property}] = {value}".format(
+            element=self.element_js,
+            property=escape_js(property_name),
+            value=escape_js(property_value)
+        ))
+
+    def node_method(self, method_name):
+        """ Return function which will call the specified method of the element """
+
+        @empty_strings_as_none
+        def call(*args):
+            result = self.tab.evaljs(u"{element}[{method}]({args})".format(
+                element=self.element_js,
+                method=escape_js(method_name),
+                args=','.join([arg.element_js if isinstance(arg, HTMLElement) else escape_js(arg) for arg in args])
+            ))
+            return self.return_html_element_if_node(result)
+
+        return call
+
+    def mouse_click(self, x=0, y=0, button="left"):
+        """ Click on the element """
+        self.assert_element_exists()
+        dimensions = self.tab.evaljs(
+            DIMENSIONS_JS_FUNC % self.element_js,
+        )
+
+        in_bounds = 0 <= x < dimensions["width"] and 0 <= y < dimensions["height"]
```

Comment:
Is this validation needed? Maybe we should allow user to click everywhere he wants? For example, I want to click on the element which is located above of my element and I will pass negative `y` to this method.